Add Solr queries to collection plugin

Erikmitk opened the pull request:
This pull request adds Solr queries to the collection plugin. Since #261 allows to define virtual collections via a Solr query those collections should be able to be shown in the collection plugin.

To be able to query the Solr an additional field for selecting a Solr core has been added to the plugin configuration.

I cleaned up some code and replaced several database queries for the corresponding Solr queries to get the necessary data.

**CollectionList**
1. fetch information about selected collections from database
2. for each collection
-- query titles and volumes info from Solr
-- generate output

**SingleCollection**
1. fetch collection information from DB
2. fetch corresponding document UIDs from Solr
3. fetch info from those documents from DB
4. generate output

Step 3 could be eliminated. But we would have to make additional changes. This particular query contains the database field metadata_sorting. That field includes all information for sorting the list view if I got that right. I haven’t figured out how the sorting actually works so that might be wrong.

![image](https://user-images.githubusercontent.com/180686/36263188-ea502f08-1269-11e8-88ea-337c30de5bec.png)

We could probably fetch the values for sorting the documents from the index, but with the options in the selector we would have to add `author` and `year` to the Solr schema. At the moment they are only dynamic fields.

frank-ulrich-weber committed:
Removes deprecated function array_search_recursive

frank-ulrich-weber committed:
Query only the required index fields

frank-ulrich-weber committed:
Avoid multiple function calls and calculations

frank-ulrich-weber committed:
*Move Highlighting from tx_dlf_search to tx_dlf_list

frank-ulrich-weber committed:
Use FastVectorHighlighter

frank-ulrich-weber committed:
Call parseTS only once per metadata

frank-ulrich-weber committed:
*Only the facet values are required and not the search result.

We don't need to fetch the search results, because only the facet values
are required. The number of facet values is limited by facet.limit
already.

frank-ulrich-weber committed:
Fix: Initialize $params array properly and use $params for search

frank-ulrich-weber committed:
Fixes some indention issues

frank-ulrich-weber committed:
Better readability by avoiding short names (Codacy)

Erikmitk committed:
Add Solr core to plugin configuration

Erikmitk committed:
Add Solr core to plugin configuration

Erikmitk committed:
Replace database queries with Solr queries where possible

Erikmitk committed:
Replace database queries with Solr queries where possible

BFallert committed:
Change url UB Mannheim

BFallert committed:
change UB Mannheim to https

sebastian-meyer committed:
Merge pull request #269 from BFallert/url_update

Url update

frank-ulrich-weber committed:
Avoids empty fields in the index by using empty default values.

sebastian-meyer committed:
Fixed comment for clarity.

sebastian-meyer committed:
Merge pull request #270 from frank-ulrich-weber/emptyFacetsFix

Avoids empty fields in the index by using empty default values.

sebastian-meyer committed:
Fixes quoting style in tx_dlf_list

sebastian-meyer committed:
Fixes quoting style in tx_dlf_solr

sebastian-meyer committed:
Fixes PHPDoc in tx_dlf_listview

sebastian-meyer committed:
Merge pull request #259 from frank-ulrich-weber/performaceOptimizationsNewStructure

Multiple performance optimizations

sebastian-meyer commented on plugins/collection/class.tx_dlf_collection.php:
```diff
@@ -95,53 +95,33 @@ public function main($content, $conf) {
-
-            } else {
-
-                $additionalWhere .= ' AND NOT tx_dlf_collections.fe_cruser_id=0';
```
Your simplification misses this case ($this->conf['show_userdefined'] > 0 && empty($GLOBALS['TSFE']->fe_user->user['uid']). This case is important, because it is used to show all userdefined collections.

sebastian-meyer commented on plugins/collection/flexform.xml:
```diff
@@ -57,6 +57,21 @@
+						<TCEforms>
+							<displayCond>FIELD:pages:REQ:true</displayCond>
+							<exclude>1</exclude>
+							<label>LLL:EXT:dlf/plugins/oai/locallang.xml:tt_content.pi_flexform.solrcore</label>
```
This is the flexform for the collection plugin, but you are referencing the language file of the oai plugin. Please try to avoid cross-plugin dependencies!

sebastian-meyer commented on plugins/collection/flexform.xml:
```diff
@@ -57,6 +57,21 @@
+								<itemsProcFunc>tx_dlf_tceforms->itemsProcFunc_solrList</itemsProcFunc>
+								<size>1</size>
+								<maxitems>1</maxitems>
+								<minitems>0</minitems>
```
I am not sure, but shouldn't this be a mandatory field (i.e. minitems=1)?

sebastian-meyer committed:
Always use toplevel collection for indexing

sebastian-meyer committed:
Rename goobi -> kitodo

sebastian-meyer commented on plugins/collection/class.tx_dlf_collection.php:
```diff
@@ -95,53 +95,33 @@ public function main($content, $conf) {
      * @return	string		The list of collections ready to output
      */
     protected function showCollectionList() {
-
```
Please preserve those empty lines throughout the file (there are many occasions where you removed them).

sebastian-meyer commented on plugins/collection/class.tx_dlf_collection.php:
```diff
@@ -154,135 +134,130 @@ protected function showCollectionList() {
+            $volumes = $titles;
 
-                } else {
+            // All titles except those that are part of something
```
You confused titles and volumes here. Titles are all documents which do not have an ancestor (i.e. partof == 0), but volumes are all documents which either have an ancestor (i.e. partof != 0) or no descendant (i.e. no other documents exist with descendant.partof == ancestor.uid).
The way you calculate titles and volumes always results in less volumes than titles, but in reality that can never be the case. The number of volumes must be at least as big as the number of titles, almost always it is even bigger.

sebastian-meyer reviewed (changes requested):
There are some inconsistencies and errors, please check my inline comments. And please have a look at manually created collections to avoid running into the same bug as with the OAI plugin.

Commit:
add l10n_mode settings to collections fields

See
https://docs.typo3.org/typo3cms/TCAReference/7.6/Reference/Columns/Index.html#l10n-mode
for background.

l10n_mode = exclude --> the record isn't shown in translated language

l10n_mode = mergeIfNotBlank --> the value of default language is shown
if field is empty in translated language record

label is required --> no mergeIfNotBlank is possible

Commit:
only the collection record with default language may be used to get index_name

Commit:
the indexing backend module may show only records of default language

Commit:
get the language record overlay in collections plugin

This is done the same way as in metadata plugin.

Things to know:

* the record to overlay must include uid,pid and sys_lanugage_uid.
* the fields to overlay may not be changed. tx_dlf_collections.label AS
 collLabel is forbidden!

lutzhelm committed:
Fix #276 - TOC plugin flexform should not require a template

Commit:
volumes may only be counted from collection in default language

sebastian-meyer committed:
Merge pull request #274 from albig/fix-multilang-collection

Fix #273 multilanguage support of collection records

sebastian-meyer committed:
Merge pull request #277 from lutzhelm/fix-276

Fix #276 - TOC plugin flexform should not require a template

sebastian-meyer committed:
Add version requirements to readme

fixes #275

sebastian-meyer committed:
Try to match documents by location

fixes #272

sebastian-meyer committed:
Add collection information to physical structures

lutzhelm committed:
Fix #279 - IIIF info.json property "tiles" is optional

Added tilesize 256x256 and scaleFactors 1, 2, 4, 8, 16 as default values
without checking the info.json profile if that is supported because
dlfViewerSource.IIIF requires compliance level 1 or 2 anyway.

sebastian-meyer committed:
Fix language overlay for collections

sebastian-meyer committed:
Merge pull request #280 from lutzhelm/fix-279-iiif-optional-tiles

Fix #279 - IIIF info.json property "tiles" is optional

sebastian-meyer committed:
Merge pull request #281 from kitodo/fix-collections

Fix language overlay for collections

sebastian-meyer committed:
Add collections from database to document metadata

sebastian-meyer committed:
Fix some collection queries

Commit:
compare value with simple array

The comparision is done at this point with the following values/arrays:

$resArray['uid'] = 3123;

$record['subparts'][0]['u'] = 4123;
$record['subparts'][1]['u'] = 2123;
$record['subparts'][2]['u'] = 3123;

array_search(3123, $record['subparts']) --> no result

Solutions:

array_search(3123, array_column($record['subparts'], 'u') --> 2

or

array_search(array('u' => 3123), $record['subparts']) --> 2

Commit:
take solution 2 for array_search as solution 1 does not always work

sebastian-meyer committed:
Remove unused variable

Codacy fix

sebastian-meyer committed:
Match location and recordId

sebastian-meyer committed:
Merge pull request #284 from albig/fix-283

fix #283: compare value with simple array

sebastian-meyer committed:
Add composer.json

sebastian-meyer committed:
Fix composer.json

sebastian-meyer committed:
Add possibility to use different fulltext formats

sebastian-meyer committed:
Fix composer.json

typo3/cms-core is onyl provided for TYPO3 8.7+

sebastian-meyer committed:
Merge pull request #278 from kitodo/fix-272

Try to match documents by location

sebastian-meyer committed:
Merge pull request #285 from kitodo/fix-197

Add possibility to use different fulltext formats

Commit:
OAI solr query should always respect toplevel setting

sebastian-meyer committed:
Merge branch 'master' into fix-271

sebastian-meyer committed:
Merge pull request #287 from albig/fix-286

OAI solr query should always respect toplevel setting

sebastian-meyer committed:
Merge pull request #282 from kitodo/fix-271

Fix using manually created collections as OAI sets

Commit:
make solr limit configurable in OAI plugin

sebastian-meyer committed:
Fix typo in locallang.xml

sebastian-meyer committed:
Merge pull request #289 from albig/fix-288

fix #288: make solr limit configurable in OAI plugin

sebastian-meyer committed:
Cast structure ID to string

Erikmitk committed:
Add missing case to $showUserDefinedCollections

Erikmitk committed:
Add missing case to $showUserDefinedCollections

Erikmitk committed:
Fix process of determining volumes

Erikmitk committed:
Fix process of determining volumes

Erikmitk committed:
Add changes that got removed by rebasing

Erikmitk committed:
Merge branch 'solr_collection_extension' of https://github.com/Erikmitk/kitodo-presentation into solr_collection_extension

# Conflicts:
#	plugins/collection/class.tx_dlf_collection.php

Erikmitk committed:
Reference correct lamguage file

Erikmitk committed:
Make Solr Core a mandatory field

Since we have to ask the Solr it has to be specified which core to use.

Erikmitk committed:
Re-added empty lines

Erikmitk commented:
I am gonna open up a new request for this, because there were a lot of changes in the mean time and I had to rebase my branch. This PR would be quite confusing.
